### PR TITLE
Deploy and Upgrade contract API on Arwen

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

**Review by camilbancioiu — CHANGES_REQUESTED** [review]
Only requirement: `upgradeContract()` has been changed on `master`. Make sure the new `upgradeFromSourceContract()` is also updated in the same way.

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -1421,6 +1433,141 @@ func v1_3_upgradeContract(
...
+	valueOffset int32,
+	sourceContractAddressOffset int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) {
+	runtime := host.Runtime()
+	metering := host.Metering()
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.CreateContract
```
[the remark this answers is not in the document]
Comment (matei-p) [reply arwen/elrondapi/elrondei.go:1446]: No

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -2761,19 +2724,9 @@ func v1_3_createContract(
...
-			Arguments:   data,
-			CallValue:   big.NewInt(0).SetBytes(value),
-			GasPrice:    0,
-			GasProvided: metering.BoundGasLimit(gasLimit),
-		},
-		ContractCode:         code,
-		ContractCodeMetadata: codeMetadata,
-	}
+	valueAsInt := big.NewInt(0).SetBytes(value)
+	newAddress, err := createContractCall(sender, data, valueAsInt, metering, gasLimit, code, codeMetadata, host, runtime)
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:2728]: Consider renaming to `createContract()`

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -1390,6 +1402,104 @@ func v1_3_upgradeContract(
 	gasToUse = math.MulUint64(gasSchedule.BaseOperationCost.DataCopyPerByte, uint64(length))
 	metering.UseGas(gasToUse)
 
+	upgradeContract(host, calledSCAddress, code, codeMetadata, value, data, gasLimit)
+}
+
+//export v1_3_upgradeFromSourceContract
+func v1_3_upgradeFromSourceContract(
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:1409]: This API function must also take the code metadata, just like `upgradeContract` does.

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -2639,6 +2739,121 @@ func v1_3_createContract(
...
+func v1_3_deployFromSourceContract(
+	context unsafe.Pointer,
+	gasLimit int64,
+	valueOffset int32,
+	sourceContractAddressOffset int32,
+	resultAddressOffset int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:2752]: Must also take `codeMetadata` as parameter, regardless of what the source contract has.

File: arwen/elrondapi/elrondei.go [hunk 5]
```diff
@@ -2639,6 +2739,121 @@ func v1_3_createContract(
...
+	code, err := blockchain.GetCode(sourceContractAddress)
+	if arwen.WithFaultAndHost(host, err, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return nil, err
+	}
+
+	contract, err := blockchain.GetUserAccount(sourceContractAddress)
+	if arwen.WithFaultAndHost(host, err, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return nil, err
+	}
+	codeMetadata := contract.GetCodeMetadata()
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:2826]: Use metadata provided as parameter at runtime instead.

File: arwen/hosttest/execution_test.go [hunk 6]
```diff
@@ -2237,6 +2237,77 @@ func TestExecution_CreateNewContract_Success(t *testing.T) {
...
+					// returned by the deployer contract
+					[]byte("succ"),
+				)
+		})
+}
+
+func TestExecution_UpgradeContractFromExistingCode_Success(t *testing.T) {
+
+	arwen.SetLoggingForTests()
+
```
Comment (camilbancioiu) [inline arwen/hosttest/execution_test.go:2277]: Forgotten logging setup.

File: arwen/elrondapi/elrondei.go [hunk 7]
```diff
@@ -1390,6 +1402,104 @@ func v1_3_upgradeContract(
...
+}
+
+// UpgradeFromSourceContractWithTypedArgs - upgradeFromSourceContract with args already read from memory
+func UpgradeFromSourceContractWithTypedArgs(
+	host arwen.VMHost,
+	sourceContractAddress []byte,
+	destCountractAddress []byte,
+	value []byte,
+	data [][]byte,
+	gasLimit int64,
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:1472]: Must take `codeMetadata` as parameter, coming from the caller contract.

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**